### ISSUE TITLE
simultaneous playback of multiple streams with MUX component not working with latest sof FW (20-march-2023)

### ISSUE THREAD

**Issue by maruthi-machani** [thesofproject/sof#7308]
With latest sof FW (20-march-2023), simultaneous playbacks of 2 input streams using mux component only single stream is working. no audio playback observed on the second stream. Earlier sof FW (02-March-2023) simultaneous playback was working properly.

**Comment by plbossart** [comment]
"simultaneous playbacks of 2 input streams"

You already lost me here @maruthi-machani, no idea how you managed to have 'playback' and 'input streams' in the same sentend.

Please provide clearer descriptions of 
- what platform
- what topology
- what test
you are running.

**Comment by maruthi-machani** [comment]
with AMD platform, when simultaneous two input streams are played using mux component only single input stream is played. this issue is observed with the latest sof FW.
Topology uses mux component with two input streams. Attaching topology for reference.
[sof-acp-rmb-tdm8ch-dmic4ch_m4.txt](https://github.com/thesofproject/sof/files/11046953/sof-acp-rmb-tdm8ch-dmic4ch_m4.txt)
[pipe-echo-ref-dai-playback_m4.txt](https://github.com/thesofproject/sof/files/11046954/pipe-echo-ref-dai-playback_m4.txt)

**Comment by dbaluta** [comment]
@maruthi-machani I bet the problem start at this commit:

commit c399624fcfb7dfaf75d69671bc2a91ec9812a6ad
Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
Date:   Tue Dec 20 18:48:10 2022 +0200

    Audio: Mux: Convert mux and demux to module adapter



We have similar issue but with MIXER.

**Comment by plbossart** [comment]
> **Usecase**: simultaneous playback of two input streams on headset and speaker.

Did  you mean: playback of one stream on both headset and speaker? 

It's very hard to parse your explanations @maruthi-machani

**Comment by maruthi-machani** [comment]
simultaneously playing two different streams one on headset and other on speaker.

**Comment by plbossart** [comment]
Never mind, it's described in the topology where a TDM interface is used.

````
# PCM0 ----> buffer ------+
#                         |----> muxdemux ----> buffer ---->  ACPSP_play
# PCM1 ----> buffer ----->
````

**Comment by plbossart** [comment]
@maruthi-machani this isn't a topology that's used on the Intel devices, headset and amplifiers use separate I2S links.

**Comment by maruthi-machani** [comment]
@plbossart irrespective of intel/amd platforms the mux component should support TDM of multiple input streams. this specific issue is observed recently. earlier (before march 6 sof) this issue is not observed.

**Comment by bhiregoudar** [comment]
Hi Liam,
Please assign this issue to proper author

**Comment by lgirdwood** [comment]
@dbaluta @bhiregoudar we dont have any working Intel IPC3/xtos configs in main now, but I think we can probably shed some light on this via the ALSA plugin (in the lrg/topic/plugin) branch. i.e. we can run the topology on host PC.

I think in the short term we can get the plugin running in CI in IPC3 mode, but we will probably need some help from NXP/AMD to do this. ALSA plugin would give us some IPC3 & IPC4 coverage on CI. Ideally we would have a UT for each module/component type.

**Comment by maruthi-machani** [comment]
@lgirdwood could you please let us know what kind of support you need from AMD.

**Comment by lgirdwood** [comment]
@maruthi-machani apologies, was away. We may need AMD help next week to test. I going to ask @lyakh to deep dive. In the mean time can AMD provide latest FW and kernel logs (inc IPC messages) of failure. Thanks.

**Comment by maruthi-machani** [comment]
@lgirdwood Attaching the FW and kernel logs (inc IPC messages) during failure.

[debuglog.txt](https://github.com/thesofproject/sof/files/11732821/debuglog.txt)
[dmesg.log](https://github.com/thesofproject/sof/files/11732822/dmesg.log)

**Comment by maruthi-machani** [comment]
@lyakh , failure is observed with patch c399624fcfb7dfaf75d69671bc2a91ec9812a6ad. by reverting the patch failure not observed.

![image](https://github.com/thesofproject/sof/assets/101189005/d00afdb0-f7ef-4068-9f6a-c5143ca1ff2d)

**Comment by singalsu** [comment]
dmesg.txt shows "oading topology:amd/sof-tplg/sof-rmb-rt5682s-rt1019.tplg" but it does not seem to contain a mux:

![sof-rmb-rt5682s-rt1019](https://github.com/thesofproject/sof/assets/28754793/6cda1968-b936-4db3-8d2f-55a815c9bd2f)

**Comment by maruthi-machani** [comment]
we rename the topology "sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.tplg" to "sof-rmb-rt5682s-rt1019.tplg" when placing in the target.
mux initialization from above debuglog.txt

![image](https://github.com/thesofproject/sof/assets/101189005/4a37e2b1-01d1-450a-9334-a583d0cb9df5)

**Comment by singalsu** [comment]
@maruthi-machani can you please try if draft fix https://github.com/thesofproject/sof/pull/7811 works for you?

**Comment by maruthi-machani** [comment]
with above fix, simultaneous playback of two streams with MUX component is working fine.
